### PR TITLE
Feature/bootstrap behat

### PR DESCRIPTION
**Pull request by defrag**
I am not sure if u need it, but here it is :)

### CODE REVIEW

**Comment by defrag** [comment]
I also pushed fix for #6 and #2, but tommorow in case im leaving for few day to fix something.

File: Features/Context/FeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
+<?php
```
Comment (dunglas) [inline Features/Context/FeatureContext.php]: Can you add the following license prefix please?

```php
/*
 * (c) Kévin Dunglas <dunglas@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */
```

File: Features/Context/FeatureContext.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace Dunglas\AngularCsrfBundle\Features\Context;
+
+use Behat\Behat\Context\SnippetAcceptingContext;
+use Dunglas\AngularCsrfBundle\Features\Context\Fixtures\TestKernel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+class FeatureContext implements SnippetAcceptingContext
```
Comment (dunglas) [inline Features/Context/FeatureContext.php]: Can you add a DocBlock with a `@author` tag?

File: Features/Context/FeatureContext.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     */
+    public function iSendPostRequestToWithInvalidCsrfHeader($path)
+    {
+        $request = Request::create($path, 'POST');
+        $request->headers->add(array(
+            'X-XSRF-TOKEN' => 'invalid_csrf_token'
+        ));
+
+        $this->lastResponse = $this->app->handle($request);
+
```
Comment (dunglas) [inline Features/Context/FeatureContext.php]: Please remove this line.

File: Features/Context/FeatureContext.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /**
+     * @Then the response code should be :arg1
+     */
+    public function theResponseShouldCodeBe($code)
+    {
+        if ($this->lastResponse->getStatusCode() !== (int) $code) {
```
Comment (dunglas) [inline Features/Context/FeatureContext.php]: A `@Transform` method should better fit: http://docs.behat.org/en/latest/guides/2.definitions.html#step-argument-transformations

File: Features/Context/FeatureContext.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            throw new \RuntimeException(sprintf("Response code expected %s, but got %s", $code, $this->lastResponse->getStatusCode()));
```
Comment (dunglas) [inline Features/Context/FeatureContext.php]: Please use simple quotes (consistency).

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<?php
+
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Same here for the license header and the DockBlock with `@author`.

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+use Dunglas\AngularCsrfBundle\Features\Context\Fixtures\TestBundle\Form\Type\CsrfProtectedType;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class TestController extends Controller
+{
+    public function createAction()
+    {
+        return new Response("Success", 201);
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Same for quotes.

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
+    public function csrfProtectedAction(Request $request)
+    {
+        $form = $this->createForm(new CsrfProtectedType());
+        $form->submit($request->request->all());
+
+        if ($form->isValid()) {
+            return new Response("Success", 200);
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Same for quotes.

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        return new Response("Success", 400);
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Same for quotes.

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Please remove the empty line.

File: Features/Context/Fixtures/TestBundle/Controller/TestController.php [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    }
+}
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Controller/TestController.php]: Please add an empty line at the end of every file.

File: Features/Context/Fixtures/TestBundle/Form/Type/CsrfProtectedType.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
+<?php
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/Form/Type/CsrfProtectedType.php]: Header and docblock.

File: Features/Context/Fixtures/TestBundle/TestBundle.php [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<?php
```
Comment (dunglas) [inline Features/Context/Fixtures/TestBundle/TestBundle.php]: Header, docblock and empty line at the end.

File: Features/Context/Fixtures/TestKernel.php [hunk 14]
```diff
@@ -0,0 +1,59 @@
+<?php
```
Comment (dunglas) [inline Features/Context/Fixtures/TestKernel.php]: Header docblock and empty line.

File: Features/Context/Fixtures/config/default.yml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+dunglas_angular_csrf:
+    header:
+        name: X-XSRF-TOKEN
+    cookie:
+        set_on:
+            - { path: ^/$ }
+    secure:
+        - { path: ^/resource, methods: [POST, PUT, PATCH, LINK] }
+        - { path: ^/protected-resource, methods: [POST, PUT, PATCH, LINK] }
```
Comment (dunglas) [inline Features/Context/Fixtures/config/default.yml]: empty line.

File: Features/securing_endpoints.feature [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    Given I send POST request to "/resource" with invalid csrf header
+    Then the response code should be 403
+
+  Scenario: The form csrf token is disabled in favor of header one
+    Given I send POST request to "/protected-resource" with valid csrf header
+    Then the response code should be 200
+
+  Scenario: The form csrf token is not disabled in favor of header one if its invalid
+    Given I send POST request to "/protected-resource" with invalid csrf header
+    Then the response code should be 403
```
Comment (dunglas) [inline Features/securing_endpoints.feature]: Empty line.

File: Form/Extension/DisableCsrfExtension.php [hunk 17]
```diff
@@ -0,0 +1,81 @@
+<?php
```
Comment (dunglas) [inline Form/Extension/DisableCsrfExtension.php]: Header docblock and empty line.

File: behat.yml.dist [hunk 18]
```diff
@@ -0,0 +1,8 @@
+default:
+    autoload:
+        '': %paths.base%/features/bootstrsap
+    suites:
+        default:
+            paths:    [ %paths.base%/features ]
+            contexts:
+                - Dunglas\AngularCsrfBundle\Features\Context\FeatureContext
```
Comment (dunglas) [inline behat.yml.dist]: Empty line.

File: behat.yml.dist [hunk 19]
```diff
@@ -0,0 +1,8 @@
+default:
```
Comment (dunglas) [inline behat.yml.dist]: Why `.dist`?

File: composer.json [hunk 20]
```diff
@@ -17,13 +17,20 @@
         "symfony/config": "~2.4",
         "symfony/http-foundation": "~2.4",
         "symfony/http-kernel": "~2.4",
-        "symfony/security-csrf": "~2.4"
+        "symfony/security-csrf": "~2.4",
+        "behat/behat": "3.0.*"
```
Comment (dunglas) [inline composer.json]: Behat must be a `require-dev` dependency IMO.

Comment (dunglas) [inline composer.json]: And prefer the `~3.0` (or better `^3.0`) notation.

File: composer.json [hunk 21]
```diff
@@ -32,4 +39,3 @@
         }
     }
 }
-
```
Comment (dunglas) [inline composer.json]: Keep the empty line please.

**Comment by dunglas** [comment]
Thanks for your awesome work @defrag. I've added some (minor) comments.
Behat tests was a need!

**Comment by defrag** [comment]
@dunglas I fixed most of the things from comments. The rest i can do next week, unless u want to fix them yourself :P Cheers.

**Comment by dunglas** [comment]
Thanks! It can wait next week (I'm very busy for now).

**Comment by defrag** [comment]
@dunglas ive pushed cs fixes u asked for.

**Comment by dunglas** [comment]
Thanks @defrag

**Comment by dunglas** [comment]
@defrag can you add some lines in the README about the form extension? Then I'll release a new version of the bundle.

**Pull request merged**